Update ROMClass walk to support JITServer ROMClass serialization on JDK 11+

AlexeyKhrabrov opened the pull request:
This pull request fixes the following issues in the ROMClass walk implementation that prevented ROMClass serialization in JITServer (#11331) from working correctly.

- Constant dynamic CP entries were not visited.
- Nest members were not visited.
- Source debug extension was erroneously handled as a `J9UTF8` string while it is actually stored as a `J9SourceDebugExtension` structure.

The DDR ROMClass walker is also updated to reflect (most of) these changes. Note that DDR currently doesn't seem to handle any struct fields that are only present in JDK 11+ (such as `nestHost` and `nestMembers` in `J9ROMClass`), but fixing that would be out of scope of this PR.

AlexeyKhrabrov commented:
@tajila This is another set of changes required for ROMClass serialization in JITServer, hopefully the last one. Can I please get a review? @mpirvu FYI.

tajila commented on runtime/util/romclasswalk.c:
```diff
@@ -693,14 +709,20 @@ static void allSlotsInConstantPoolDo(J9ROMClass* romClass, J9ROMClassWalkCallbac
 			default:
-				/* unknown cp type - bail */
-				return;
+				/* unknown cp type */
```
This is an error path, did you actually detect this path being taken?

AlexeyKhrabrov replied:
Not sure; I would have to check that specifically. It definitely would be taken for any class that has constant dynamic CP entries. Now that all the CP entry types are handled properly, this path should never be taken. If in the future new CP entry types are added but not handled by this code, I think it's better to visit them here as "unknown" rather than silently skipping the rest of the constant pool.

tajila replied:
Okay, my preference would be to add an assertion so we know there is a problem.

tajila replied:
`Assert_Util_unreachable`

AlexeyKhrabrov replied:
Sure. I'll make the changes.

tajila commented:
@fengxue-IS please review these changes. Also were new CP types introduced for the anonClass patching work?

fengxue-IS reviewed (approved):
lgtm
@tajila the CP patching for anonClass doesn't use any new CP_TYPE

AlexeyKhrabrov commented:
@tajila I've updated the code to assert on unknown CP entry types. I've also changed the DDR walker to throw an exception in this case (so that it's more consistent with VM code).

tajila commented:
Jenkins test sanity win jdk8

tajila commented:
Jenkins test sanity,extended plinux jdk11

tajila commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/RomClassWalker.java:
```diff
@@ -631,31 +636,41 @@ private void allSlotsInConstantPoolDo() throws CorruptDataException
 				classWalkerCallback.addSlot(clazz, SlotType.J9_SRPNAS, ref.nameAndSignatureEA(), "cpFieldNAS");
 				classWalkerCallback.addSlot(clazz, SlotType.J9_U32, ref.classRefCPIndexEA(), "cpFieldClassRef");
-				
+
```
looks like we are missing `J9CPTYPE_INTERFACE_STATIC_METHOD` and `J9CPTYPE_INTERFACE_INSTANCE_METHOD` here

AlexeyKhrabrov replied:
Fixed in the commit I've just added

tajila commented:
Jenkins test sanity,extended plinux jdk11

keithc-ca commented on runtime/ddr/vmddrstructs.properties:
```diff
@@ -238,6 +238,8 @@ ddrblob.typeoverride.J9ROMNameAndSignature.signature=J9SRP(struct J9UTF8)
 
 ddrblob.typeoverride.J9ROMMethodTypeRef.signature=J9SRP(struct J9UTF8)
 
+ddrblob.typeoverride.J9ROMConstantDynamicRef.nameAndSignature=J9SRP(struct J9ROMNameAndSignature)
```
Should this not also be added to `runtime/ddr/overrides-vm`?

tajila replied:
Yes, this was missed. I've created https://github.com/eclipse/openj9/pull/11706 to address this.
